crates/e2e/src/contract_results.rs

```diff
@@ -0,0 +1,204 @@
+use ink::codegen::ContractCallBuilder;
+use ink_env::{
+    call::FromAccountId,
+    Environment,
+};
+use ink_primitives::MessageResult;
+use pallet_contracts_primitives::{
+    CodeUploadResult,
+    ContractExecResult,
+    ContractInstantiateResult,
+    ExecReturnValue,
+};
+use std::{
+    fmt::Debug,
+    marker::PhantomData,
+};
+
+/// Result of a contract instantiation.
+pub struct InstantiationResult<E: Environment, EventLog> {
+    /// The account id at which the contract was instantiated.
+    pub account_id: E::AccountId,
+    /// The result of the dry run, contains debug messages
+    /// if there were any.
+    pub dry_run: ContractInstantiateResult<E::AccountId, E::Balance, ()>,
+    /// Events that happened with the contract instantiation.
+    pub events: EventLog,
```

Comment:
So the concrete type for this is `ExtrinsicEvents`, what's the idea for the alternative backend?

Certainly we would want to access the events with a consistent API regardless of the backend being used.

Reply:
for 100% drink will support events (it's in progress), but it's not there yet; once it arrives, I will put a common trait over `EventLog` here for both backends; I think it should be at least iterable over some `Event` type with lookup